.github/workflows/check-upstream.main.kts

```diff
@@ -4,33 +4,100 @@
 
 @file:Repository("https://bindings.krzeminski.it/")
 @file:DependsOn("actions:checkout:v4")
+@file:OptIn(ExperimentalKotlinLogicStep::class)
 
 import io.github.typesafegithub.workflows.actions.actions.Checkout
+import io.github.typesafegithub.workflows.actions.actions.Checkout.FetchDepth.Infinite
+import io.github.typesafegithub.workflows.annotations.ExperimentalKotlinLogicStep
 import io.github.typesafegithub.workflows.domain.RunnerType
+import io.github.typesafegithub.workflows.domain.triggers.Cron
+import io.github.typesafegithub.workflows.domain.triggers.Push
+import io.github.typesafegithub.workflows.domain.triggers.Schedule
 import io.github.typesafegithub.workflows.domain.triggers.WorkflowDispatch
 import io.github.typesafegithub.workflows.dsl.workflow
+import kotlin.io.path.Path
+import kotlin.io.path.readText
+import kotlin.io.path.writeText
+
+val badgeFileName = "commits-to-upstream-badge.svg"
 
 workflow(
     name = "Check upstream",
     on = listOf(
         WorkflowDispatch(),
+        Push(branches = listOf("main")),
+        Schedule(triggers = listOf(Cron(minute = "0", hour = "0", dayWeek = "6"))), // Once a week.
     ),
     sourceFile = __FILE__,
 ) {
     job(
         id = "check",
         runsOn = RunnerType.UbuntuLatest,
     ) {
-        uses(action = Checkout())
+        uses(action = Checkout(fetchDepth = Infinite))
```

Comment:
note: I think it might not work. By default, as far as I know, the token used for checkout has read-only permissions. You need to request write permissions for the repo

Reply:
question: What do you think if instead of checking out the main branch we checkout `commits-to-upstream-badge` branch right away? `ref` parameter should allow that. In this case, you won't probably need all the staff you do with `stash` command when committing changes.

Regarding the `upstream-commit.txt` I think you can check out the main branch for SnakeKMP repo just like you checked out the `snakeyaml-engine` repo.
```
git clone --branch master --single-branch https://bitbucket.org/snakeyaml/snakeyaml-engine.git
git clone --branch ${{ github.ref }} --single-branch ${{ github.repositoryUrl }} SnakeKMP
cd snakeyaml-engine
git log --oneline $(cat ../SnakeKMP/upstream-commit.txt)..master | wc -l > ../number-of-commits.txt
```

Reply:
> I think it might not work

I tested it by triggering the workflow manually, works: https://github.com/krzema12/snakeyaml-engine-kmp/actions/runs/11724730307. I know what you mean. I tried looking up info about the default but it's somehow not straightforward.

Regarding checking out the target branch, yeah - I had it like this before, but then I used the Kotlin-based step where github-workflows-kt has to have the script checked out, hence such complication. I think having the step of building the SVG in Kotlin is not worth the hassle of later operarions on git. I'll rewrite this step to Bash and then everything should become simpler like you suggest.

Reply:
About rewriting step to bash: have you considered using [shields.io](https://shields.io/docs/static-badges)? In this case it would be just a single curl or wget request. This of course introduces dependency to that service but it will be much clearer than having svg defined as a string (in my opinion)

Reply:
I haven't, it's a great idea, thanks!